allennlp/data/fields/copy_map_field.py

```diff
@@ -0,0 +1,51 @@
+from typing import Dict, List
+
+from overrides import overrides
+import torch
+
+from allennlp.common.util import pad_sequence_to_length
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.data.tokenizers.token import Token
+from allennlp.data.fields.field import Field
+
+
+class CopyMapField(Field[torch.Tensor]):
+    """
+    A `CopyMapField` is used to map tokens in one namespace to tokens in another namespace.
+    It is used by seq2seq models with a copy mechanism that copies tokens from the source
+    sentence into the target sentence.
+
+    Parameters
+    ----------
+    source_tokens : ``List[Token]``
+        The tokens from the source sentence.
+    target_namespace : ``str``
+        The namespace that the tokens from the source sentence will be mapped to.
+    """
+
+    def __init__(self,
+                 source_tokens: List[Token],
+                 target_namespace: str) -> None:
+        self._source_tokens = source_tokens
+        self._target_namespace = target_namespace
+        self._mapping_array: List[int] = None
+
+    @overrides
+    def index(self, vocab: Vocabulary):
+        self._mapping_array = [vocab.get_token_index(x.text, self._target_namespace)
```

Comment:
Ah, I see, in misunderstood what was going on here on my first read.  I was thinking you used this array to determine when copying happened, but you use `source_token_ids` and `target_token_ids` for that.  It looks like this field is mostly necessary for beam search, both for computing the selective attention (when I predict a target token I need to find occurrences of it in the source sequence) and for combining generation and copy probabilities during search (so that sorting by option probability works correctly).

I spent quite a long time digging through this logic trying to figure out if there were a simpler way to pass in the copy information, but I finally convinced myself that you need all of the fields that you have, and that this is a good way to do it.  If you didn't have to worry about beam search / decoding, I think you could make things a lot simpler.  But there's not a whole lot of point to this model if you can't do decoding =).

So, it would have been nice to have some documentation describing the decisions around what fields are present, as I mentioned above.  That would have saved me a whole lot of digging through the logic.